cpp/stats_server.cc

```diff
@@ -0,0 +1,266 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <unistd.h>
+#include <cmath>
+#include <ctime>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <grpcpp/ext/proto_server_reflection_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#include "examples/protos/account.grpc.pb.h"
+#include "examples/protos/stats.grpc.pb.h"
+
+using account::Account;
+using account::GetUserInfoRequest;
+using account::GetUserInfoResponse;
+using account::MembershipType;
+using grpc::Channel;
+using grpc::ChannelArguments;
+using grpc::ClientContext;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::ServerWriter;
+using grpc::Status;
+using grpc::StatusCode;
+using stats::PriceRequest;
+using stats::PriceResponse;
+using stats::Stats;
+
+class StatsServiceImpl final : public Stats::Service {
+ public:
+  void SetAccountClientStub(std::unique_ptr<Account::Stub> stub) {
+    account_stub_ = std::move(stub);
+  }
+
+  void SetHostName(const std::string& hostname) { hostname_ = hostname; }
+
+  void SetPremiumOnly(const bool premium_only) { premium_only_ = premium_only; }
+
+ private:
+  bool ObtainAndValidateUserAndMembership(ServerContext* server_context) {
+    const std::multimap<grpc::string_ref, grpc::string_ref> metadata =
+        server_context->client_metadata();
+    for (auto iter = metadata.begin(); iter != metadata.end(); ++iter) {
+      if (iter->first == "authorization")
+        token_ = std::string(iter->second.data(), iter->second.size());
+      else if (iter->first == "membership")
+        membership_ = std::string(iter->second.data(), iter->second.size());
+    }
+    // Stats server started with --premium_only flag, but metadata says this is
+    // from a non-premium request, reject it.
+    if (premium_only_ && membership_ != "premium") {
+      std::cout << "requested membership is non-premium but his this an "
+                   "premium-only server"
+                << std::endl;
+      return false;
+    }
+    GetUserInfoRequest request;
+    request.set_token(token_);
+    ClientContext context;
+    GetUserInfoResponse response;
+    Status status = account_stub_->GetUserInfo(&context, request, &response);
+    if (status.ok()) {
+      auto metadata_hostname =
+          context.GetServerInitialMetadata().find("hostname");
+      if (metadata_hostname != context.GetServerInitialMetadata().end()) {
+        std::cout << "server host: "
+                  << std::string(metadata_hostname->second.data(),
+                                 metadata_hostname->second.length())
+                  << std::endl;
+      }
+    } else {
+      std::cout << status.error_code() << ": " << status.error_message()
+                << std::endl;
```

Comment:
Is it better for `ObtainAndValidateUserAndMembership` to return a status, so error messages like this can be sent back to the client?